### ISSUE TITLE
Binary string encoding to buffer using correct encoding type

### ISSUE THREAD

**Issue by arnas** [dougmoscrop/serverless-http#81]
Hello,  

I have been dealing with issues when trying to upload binary data when this data is converted to a binary string. Basically https://github.com/dougmoscrop/serverless-http/blob/c70957b47ac66e363c7179a6e0a4cd8c6d78a2ee/lib/request.js#L15 there is a line that if the body is a string it creates a buffer using `utf-8` encoding.  
Maybe it would be possible to detect what encoding was used or assume that it was `binary` based on the request header?
I have encountered this issue when I was using serverless-offline library as it currently encodes `multipart/form-data` payload to `binary` string. The issue is also raised there https://github.com/dherault/serverless-offline/issues/464

**Comment by dougmoscrop** [comment]
The binary detect code let's you provide a completely custom function - have you tired this? Does it not work?

**Comment by Vangel42** [comment]
Hello,

Any update on that issue ? I've got the same problem when uploading an excel file in multipart/form-data.
I don't want to use base64 as it increases file size.
Thanks and Regards

**Comment by dougmoscrop** [comment]
You have to use base64 in integration, it's how API gateway works, nothing to do with this library. But you don't *send* base64, it's just what APIg xonverrts to 

You (utf8) -> APIg (base64) -> Lambda (utf8) -> app (bodyparser, etc)

But that is unrelated to this issue, this issue is about having non-utf8 encoding of the original file.

(Side note, you should try not to send files thru Lambda, but use S3...)

**Comment by Vangel42** [comment]
Thanks for the answer.
However I'm sending binary content, as it is an excel file. It is the line that arnas pointed out that converts it to utf8 and messes it up. If I replace on that line 'utf8' by 'binary' it works well.

**Comment by dougmoscrop** [comment]
Sorry, I meant "Buffer" above (which I fixed). That buffer is either utf8 or base64 depending on how API Gateway is configured.

Changing it to `binary` works locally because, as far as I can tell, `serverless-offline` is not reproducing API Gateway behaviour properly (or they assume you won't do what you are so its kind of undefined)

You _must_ configure _binary content handling_ in API Gateway. This will invoke the Lambda function with the base64 encoded string and this library will detect the `isBase64Encoded` etc as long as your binary mime types are set.  You do not encode it in base64, API Gateway does.

None of this is about this library, if you deploy a bare Lambda function, you will see, I think.

I pushed a small, vanilla Lambda function to the integration test package in this repo, which just shows, if you do:

> curl -F 'data=@test/integration/image.png' {{url}}

it will return:

> {"actualBytes":1320948,"expectedBytes":725302,"bytesEqual":false}

This is because API Gateway is taking the binary data and doing some encoding stuff to it, in order to make it acceptable as part of the JSON payload that Lambda expects as part of the Invoke contract.  (A Lambda Invoke is just an HTTP POST with AWS signing).

Since JSON does not support binary strings, you can not send binary data to Lambda. Now, you could argue that this binary string is better than base64, that I dunno, but it's not what API Gateway suggests.

**Comment by dougmoscrop** [comment]
Seems like they're aware of the issue:

https://github.com/dherault/serverless-offline/pull/784

**Comment by dougmoscrop** [comment]
Just another example:

> echo -e '\x03\xF1' | curl -X POST --data-binary @- {{echo-url-endpoint}}

results in the following body in the Lambda:

> "body": "\u0003�\n",

I'm not super sure I am right/wrong here about the issue. Is the data sent available just needs to be transformed?

**Comment by Vangel42** [comment]
You're right it is working on AWS, so it is the combination serverless-offline + serverless-http that does not work.
